### DISCUSSION TITLE
`Model` hierarchy

### DISCUSSION THREAD

**Discussion by pmli** [pymor/pymor#1794] [Ideas]
This is the current `Model` hierarchy (arrows represent inheritance, diamonds represent composition).

```mermaid
classDiagram
    direction LR
    CacheableObject <|-- Model
    ParametricObject <|-- Model
    Model <|-- StationaryModel
    Model <|-- InstationaryModel
    Model <|-- LTIModel
    Model <|-- PHLTIModel
    Model <|-- SecondOrderModel
    Model <|-- LinearDelayModel
    Model <|-- LinearStochasticModel
    Model <|-- BilinearModel
    CacheableObject <|-- TransferFunction
    ParametricObject <|-- TransferFunction
    TransferFunction <|-- FactorizedTransferFunction
    FactorizedTransferFunction *-- LTIModel
    FactorizedTransferFunction *-- PHLTIModel
    FactorizedTransferFunction *-- SecondOrderModel
    FactorizedTransferFunction *-- LinearDelayModel
    Model <|-- BaseNeuralNetworkModel
    BaseNeuralNetworkModel <|-- NeuralNetworkInstationaryModel
    BaseNeuralNetworkModel <|-- NeuralNetworkInstationaryStatefreeOutputModel
    BaseNeuralNetworkModel <|-- NeuralNetworkModel
    BaseNeuralNetworkModel <|-- NeuralNetworkStatefreeOutputModel
    InstationaryModel <|-- BaseQuadraticHamiltonianModel
    BaseQuadraticHamiltonianModel <|-- QuadraticHamiltonianModel
    BaseQuadraticHamiltonianModel <|-- ReducedQuadraticHamiltonianModel
```

There are various open questions:

- Should `LTIModel` (and other system-theoretic `Models`) be a subclass of `InstationaryModel`?
- Should `PHLTIModel` and `SecondOrderModel` be subclasses of `LTIModel`?
- Should `TransferFunction` be a subclass of `StationaryModel` with Laplace variable as a parameter?
- Should `LTIModel` be a subclass of `BilinearModel`?

---

Here are the types of models we want to cover:

1. Time-dependent problems (linear or nonlinear, parametric or nonparametric, with or without inputs, with or without outputs):

    ```math
    \begin{align*}
      M \dot{x}(t, \mu) & = f(t, x(t, \mu), u(t), \mu), \quad x(0, \mu) = x_0(\mu), \\
      y(t, \mu) & = g(t, x(t, \mu), u(t), \mu),
    \end{align*}
    ```

    or

    ```math
    \begin{align*}
      M x(k + 1, \mu) & = f(k, x(k, \mu), u(k), \mu), \quad x(0, \mu) = x_0(\mu), \\
      y(k, \mu) & = g(k, x(k, \mu), u(k), \mu).
    \end{align*}
    ```

1. Parametric stationary problems (linear or nonlinear, with or without outputs):

    ```math
    \begin{align*}
      f(x(\mu), \mu) & = b(\mu), \\
      y(\mu) & = g(x(\mu), \mu).
    \end{align*}
    ```

1. State-free continuous/discrete-time system:

    ```math
    (t, \mu) \mapsto y(t, \mu)
    \text{ or }
    (k, \mu) \mapsto y(k, \mu)
    ```

1. State-free stationary problem:

    ```math
    \mu \mapsto y(\mu)
    ```

1. Samples of a time-dependent system:

    ```math
    (t_i, \mu_i, u(t_i), x(t_i, \mu_i), y(t_i, \mu_i)) \text{ for } i = 1, 2, \ldots, N.
    ```

1. Samples of a parametric stationary problem:

    ```math
    (\mu_i, x(\mu_i), y(\mu_i)) \text{ for } i = 1, 2, \ldots, N.
    ```

Some remarks:

- transfer functions could be treated as (state-free) parametric stationary problems with the Laplace variable $s$ as a complex-valued parameter,
- in the same way, frequency response data could be thought of as samples of a parametric stationary problem (unless derivatives are also sampled...),
- samples could consist of only state (e.g., for DMD), only output (e.g., for traditional system identification), or both state and output (e.g., for ioDMD).
- it would be useful to be able to combine different types of models, e.g., subtracting samples from a ROM to obtain errors,
- time-delay systems could be covered by including state delays $x(t - \tau, \mu)$ or input delays $u(t - \tau)$ in $f$ and $g$, and defining the initial condition over $[-\tau, 0]$,
- stochastic differential equations would have noise as an input.

**Comment by artpelling** [comment]
I would like to add the problem of including some kind of data model that contains information about the data such as the input/output dimensions/spaces and so on. My following suggestion only thinks towards SysID,  where the data are matrix-valued time/frequency samples of some system output or state and assumes that system identification in pyMOR should be handled through `reductors` and not `algorithms` or even a new module.

I see the need for a `DataModel` because it could
- serve as a `fom` in data-driven reductors
- implement some kind of basic operations such as adding or subtracting data and Fourier transforms for equidistant data
- be used to compute norms
- be convenient for plotting

A `DataModel` to me is basically just an object that wraps a tuple of `(x,y)` where `x.shape == (n,)` and `y.shape == (n, p, m)` with some more attributes such as `sampling_time` and above functionalites. The problem here is that it does not fit so well to other `Models`, because it cannot be `solved` at arbitrary values (only for the sampling times/frequencies) and this restriction would have to handled weirdly.

Say pyMOR supported discrete parameter spaces, where the parameter has to be in a finite set of discrete values. Then we could have a parametric `DataModel` with a discrete time or frequency parameter and its output defined by some predefined data. This implementation would maybe be more pyMOR-ic and closer to other `Models`. E.g. if `TransferFunction` is a subclass of `StationaryModel`, then a `DataModel` of transfer function values could sit alongside it with a discrete parameter space.

I am also intrigued by the idea of calling it a `BlackBoxModel` :thinking::thought_balloon:

**Reply by pmli** [reply]
This reminds me that we discussed a bit having generic state-free `Models`, which would cover both `TransferFunction` and `NeuralNetwork[Instationary]StatefreeOutputModel`. Then it might make sense to have something like `StatefreeOutputModel` as a superclass of `Model`, and then have transfer functions, neural networks, and data as subclasses of `StatefreeOutputModel` next to `Model`.

I was also thinking about having something like `DiscreteParameterSpace` for data-driven problems. It might be nice for `StationaryModel` to have an `l2_norm` method that accepts a `ParameterSpace`, including `DiscreteParameterSpace`.

For sys-mor, maybe it makes sense to have separate data models for time and frequency data? Something similar to `FrequencyResponseData` and `TimeResponseData` in [python-control](https://python-control.readthedocs.io/en/0.9.2/classes.html)? Maybe also a class for Markov parameter data, or would that fit under time response data (for impulse response)?

**Reply by artpelling** [reply]
> This reminds me that we discussed a bit having generic state-free `Models`, which would cover both `TransferFunction` and `NeuralNetwork[Instationary]StatefreeOutputModel`. Then it might make sense to have something like `StatefreeOutputModel` as a superclass of `Model`, and then have transfer functions, neural networks, and data as subclasses of `StatefreeOutputModel` next to `Model`.

I like that! Maybe the name is a bit clunky...

> I was also thinking about having something like `DiscreteParameterSpace` for data-driven problems. It might be nice for `StationaryModel` to have an `l2_norm` method that accepts a `ParameterSpace`, including `DiscreteParameterSpace`.

Yes it would be cool to have norms also for data.

> For sys-mor, maybe it makes sense to have separate data models for time and frequency data? Something similar to `FrequencyResponseData` and `TimeResponseData` in [python-control](https://python-control.readthedocs.io/en/0.9.2/classes.html)? Maybe also a class for Markov parameter data, or would that fit under time response data (for impulse response)?

I thought that the kind of data (time or frequency) could be inferred from the `ParameterSpace`? Not sure if that would play nicely with `solve`, though. Either way some "equidistant" sublass or with a sampling_time attribute would be nice to have. Then, data could be transformed and used for discrete-time sysid.

**Reply by lbalicki** [reply]
I like the idea of the `StatefreeOutputModel` a lot. And having some data class as a subclass of the `StatefreeOutputModel` would make a lot of sense to me. I also think having subclasses for different type of data would be useful since different functionalities might be desired in each of the cases. Going with that approach it should be easy to pass an `equidistant_data` attribute or something similar. Aside from "low-dimensional" frequency/time response data (which would be Numpy data I assume), I can also see having full state data e.g., for DMD or POD + NeuralNetwork approaches. In that case `StatefreeOutputModel` might be a bit misleading. Maybe even something like `InputOutputModel` might be more intuitive?

**Comment by sdrave** [comment]
So maybe as a first relative general remark from my side: I feel that one issue with the way `Models` are currently used in pyMOR, there seems to be little benefit of being a `Model` in the first place: While it does define some common interface method, the behavior of these method is quite arbitrary. For instance, to `solve` a `StationaryModel`, you only need a parameter and get a single vector. For time-dependent problems, you usually need an input for `solve` to make sense, and you get returned a time-series. In both cases, it could actually happen that `solve` does not work, because there is no state, and you can only compute an `output`. Overall, just knowing that something is a `Model` does not seem very useful. (Are there examples?) In particular, having some `DataModel` which is a `Model` in a formal way, does seem to make things even mushier.

Thus, I think we should reevaluate if everything needs to be a `Model`. The other way around: maybe we actually do not need a `Model` base class? Can we, instead, pin down three or four mathematically well-defined objects which can act as base classes? For instance, I could imagine an `InputOutputModel` as a base class, which has an `InputOutputModelWithState` as a child (feels like we did that before ...) Then, there would be some frequency domain class which of which all transfer functions are subclasses. If feel like the data sample classes should go separate for the sake of mathematical well-definedness?

Note: I'm not saying that this is the right approach.

**Reply by pmli** [reply]
Thinking about mathematical objects sounds good. I updated the first comment with some things I could think of.

> Overall, just knowing that something is a `Model` does not seem very useful. (Are there examples?)

I guess that a nice example would be a `Reductor` that just needs the `solve` method. If I remember correctly, I asked you about this with regard to RB methods and the issue was that the `Reductor` need to know which `ProjectionBasedReductor` to use.

**Comment by sdrave** [comment]
So I could imagine something like

```mermaid
classDiagram
    direction LR
    StationaryModel <|-- StationaryModelWithState
    IOModel <|-- IOModelWithState
```

as a base class hierarchy for all the `Models` we currently have in pyMOR (modulo names).  So the `Models` without `State` would only have an `output` function but not a `solve` function. We could add the `input` parameter only to the signature of `output` for the `IOModel`. We would have to allow zero-dimensional inputs to handle time-dependent `Models` with no input, however. Not sure how much special-casing that would mean for methods of `IOModel`. Adding an additional `OModel` does not seem like a good idea, from my perspective.

I guess `TransferFunction` could then be made a `StationaryModel` if that is useful. Not sure, how `FactorizedTransferFunction` would fit in. Also, I am not so sure of the concrete benefits in doing so, compared to having a separate class which can be converted into a `StationaryModel`.

About data samples, being special cases of the `Models` in this hierarchy I am not so sure. One question would be: which methods do we want to have, which, for instance, work with any `IOModel`. Would the same methods make sense for sample sets? As a general rule of thumb, I would strongly propose that for the redesigned `Model` hierarchy [Liskov's substitution principle](https://en.wikipedia.org/wiki/Liskov_substitution_principle) should hold.  My feeling is that data samples should be independent classes, but I am happy to be convinced of the contrary.

If people feel that these four classes are a good basis for all we currently have in pyMOR, then I'd propose as a next step to try to pin down `IOModel` as exactly as possible, i.e., define the corresponding mathematical object and define the signature and semantics of its methods.

**Reply by drittelhacker** [reply]
After the discussion in the dev meeting, where for me one take-away-message was that actually data samples are not really models, I was thinking, if it would make sense to have a parallel infrastructure for those. Basically, we want to learn, or identify, a surrogate model that can explain the data. So just like for classic `Models` we have `Reductors`, we should probably have something like a `Learner` acting on a `DataSet` very much inspired by the general philosophy of what [mlr3](https://mlr3.mlr-org.com/) is doing.

I guess that is much in line with @sdrave's feeling that data samples should be separate classes, but add the fact that the surrogates are learned rather than reduced.

**Reply by lbalicki** [reply]
I like the suggestion of @drittelhacker which would also be in line with the idea of following Liskov's substitution principle. 

Letting `TransferFunction` be a subclass of `StationaryModel` would e.g., be useful for #1756. There I could then simply use any `StationaryModel` as a `fom`. The type of the `rom` would then still depend on whatever the `fom` is (e.g., `TransferFunction` for `LTIModels` and something like `BarycentricRationalFunction` for general `StationaryModels`).

Also, what about models with no output and just a state? Would that just be a `StationaryModelWithState` where the `output` is undefined or again the whole state?

**Reply by sdrave** [reply]
I am not so sure whether I would use different terminology for the data-driven case. After all, we speak of data-driven MOR and pyMOR is a MOR library, not a machine-learning library. If the focus of a method is purely on system identification and not finding a low-order surrogate, then speaking of learning makes sense, of course.

Regarding models with no output: I would treat them as models with zero-dimensional output. In particular, that means that all methods working on a generic `IOModel` (or however it will be called) need to be aware of this possibility.

**Reply by lbalicki** [reply]
> I am not so sure whether I would use different terminology for the data-driven case. After all, we speak of data-driven MOR and pyMOR is a MOR library, not a machine-learning library. If the focus of a method is purely on system identification and not finding a low-order surrogate, then speaking of learning makes sense, of course.

Well, a lot of the data-driven MOR methods are simply based on generating data from a FOM and then using system-identification to identify some ROM. I wouldn't fully rule out the idea of having different naming conventions for these types of methods, yet, since really there is no reduction happening if someone wants to use the methods for system identification. 
> 
> Regarding models with no output: I would treat them as models with zero-dimensional output. In particular, that means that all methods working on a generic `IOModel` (or however it will be called) need to be aware of this possibility.

That sounds reasonable.
